### PR TITLE
Add structured errors to Typescript Client

### LINKED ISSUE #114: TypeScript Client improvements
Client generation functionality is awesome, but I think there are some things that would make it better:

1. Add the ability to change token without re-creating a client instance. A simple solution would be to expose setter/getter for the said token instead of providing token as a constructor argument
2. Check that token is truthy before adding auth header. I've accidentally set it to `null` and requests to not-auth paths started to fail with 401
3. Return error response as parsed json instead of a stringified error. One solution could be to override the fetch implementation altogether, e.g. by providing it as one of the constructor options

Let me know what makes sense and I can implement it.

### PR DESCRIPTION
**Pull request by DomBlack**
This commit adds structured errors to the typescript client in the
form of an `APIError` object which will be thrown incase of an
error. This makes a best attempt at recovering the structured error
as reported by an Encore API and parsing it to be easily usable from
withing a client.

This also causes the client to preemptively return throw an error if
no bearer token is present when trying to call an Authenticated API.

Fixes #114

```typescript
const client = new Client(Local)

try {
    const productListing = await client.products.List()
} catch (e: Error) {
    if (isAPIError(e)) {
        console.log(`Got an ${e.code} error with message ${e.message} and details ${e.details}`)
    } else {
       console.log(`Got error ${e.message}`)
    }
}
```

### CODE REVIEW

File: cli/internal/codegen/testdata/expected_typescript.ts [hunk 1]
```diff
@@ -385,34 +399,324 @@ class BaseClient {
...
+        typeof(err.code) === "number" &&
+        typeof(err.message) === "string" &&
+        (err.details === undefined || err.details === null || typeof(err.details) === "object")
+    )
+}
+
+/**
+ * EncoreError represents a structured error as returned from an Encore application.
+ */
+export class EncoreError extends Error {
```
Comment (eandre) [inline cli/internal/codegen/testdata/expected_typescript.ts:466]: Maybe this should be `APIError`? `EncoreError` might suggest it's an error **with** Encore or **from Encore the product/platform**, as opposed to from the *Encore application*.

File: cli/internal/codegen/testdata/expected_typescript.ts [hunk 2]
```diff
@@ -375,7 +389,7 @@ class BaseClient {
     }
 
     // callAPI is used by each generated API method to actually make the request
-    public async callAPI(method: string, path: string, body?: BodyInit, params?: CallParameters): Promise<Response> {
+    public async callAPI(method: string, path: string, requiresAuth: boolean, body?: BodyInit, params?: CallParameters): Promise<Response> {
```
Comment (eandre) [inline cli/internal/codegen/testdata/expected_typescript.ts:392]: Should this be public? If we're changing the signature it suggests it's not stable and should therefore not be public?

File: cli/internal/codegen/testdata/expected_typescript.ts [hunk 3]
```diff
@@ -385,34 +399,324 @@ class BaseClient {
...
+            try {
+                let jsonBody = await response.json()
+                if (isAPIErrorResponse(jsonBody)) {
+                    body = jsonBody
+                } else {
+                    body.message += ": " + JSON.stringify(jsonBody)
+                }
+            } catch {
+                // otherwise we just append the text to the error message
+                body.message += ": " + await response.text()
```
Comment (eandre) [inline cli/internal/codegen/testdata/expected_typescript.ts:435]: If `response.json()` fails, I don't believe we can call `response.text()` as the body has already been consumed. Maybe better to call `response.text()` and then `JSON.parse()` on that?

File: cli/internal/codegen/testdata/expected_typescript.ts (hunk 2 above)
Comment (DomBlack) [reply cli/internal/codegen/testdata/expected_typescript.ts:392]: The base client is private and can’t be accessed access of this file

**Review by ekerfelt — APPROVED** [review]

**Pull request merged**